### ISSUE TITLE
test failing: test_umath.TestComplexFunctions.test_loss_of_precision(<type 'numpy.complex256'>,) (migrated from Trac #1038)

### ISSUE THREAD

**Issue by thouis** [thouis/numpy-trac-migration#2591]
Original ticket http://projects.scipy.org/numpy/ticket/1038
Reported 2009-03-05 by atmention:stsci-sienkiew, assigned to unknown.

numpy 1.3.0.dev6545
python 2.5.1
nose 0.10.4

This test fails on:
Mac OSX 10.4 (tiger), PPC, 32 bit
Solaris 8, SPARC, 32 bit

This failure not reported on:
Mac OSX 10.4 (tiger), x86, 32 bit
Red Hat Enterprise Linux, x86, 32 bit
Red Hat Enterprise Linux, x86_64, 64 bit
# 
## FAIL: test_umath.TestComplexFunctions.test_loss_of_precision(<type 'numpy.complex256'>,)

Traceback (most recent call last):
  File "/usr/stsci/pyssgdev/2.5.1/nose/case.py", line 182, in runTest
    self.test(_self.arg)
  File "/usr/stsci/pyssgdev/2.5.1/numpy/core/tests/test_umath.py", line 455, in check_loss_of_precision
    check(x_series, 2_eps)
  File "/usr/stsci/pyssgdev/2.5.1/numpy/core/tests/test_umath.py", line 437, in check
    assert np.all(d < rtol), (x[np.argmax(d)], d.max())
AssertionError: (nan, nan)

---

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2009-03-06

The PPC has weird^H unique handling of extended precision, so we might want to mark the result on that platform as a known fail. Do you know if the SPARC hardware that failed supports extended precision?

**Comment by thouis** [comment]
Comment in Trac by atmention:stsci-sienkiew, 2009-03-06

Replying to [comment:1 charris]:

> The PPC has weird unique handling of extended precision, so we might want to mark the result on that platform as a known fail. Do you know if the SPARC hardware that failed supports extended precision? 

I don't know.  Does this tell you anything?

```
memconf:  V2.4 30-Aug-2007 http://www.4schmidts.com/unix.html
hostname: grail.stsci.edu
banner:   Sun Ultra 5/10 UPA/PCI (UltraSPARC-IIi 333MHz)
manufacturer: Sun Microsystems, Inc.
model:    Ultra-5_10
Sun development codename: Darwin/Otter (Ultra 5), Darwin/SeaLion (Ultra 10)
Sun Family Part Number: A21 (Ultra 5), A22 (Ultra 10)
Solaris 8 2/02 s28s_u7wos_08a SPARC, 64-bit kernel, SunOS 5.8
1 UltraSPARC-IIi 333MHz cpu, system freq: 111MHz
CPU Units:
========================= CPUs =========================
                    Run   Ecache   CPU    CPU
Brd  CPU   Module   MHz     MB    Impl.   Mask
---  ---  -------  -----  ------  ------  ----
 0     0     0      333     2.0   12       1.3

Status of virtual processor 0 as of: 03/06/09 09:22:01
  on-line since 03/04/09 02:29:55.
  The sparcv9 processor operates at 333 MHz,
        and has a sparcv9 floating point processor.
```

Otherwise, I can ask the machine, if you can tell me how.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2009-03-06

I googled it, the SPARC has quad precision instead of extended precision. I'll take a look at the test to see if that could be a problem. I don't see where the nans are coming from.

I don't like the assert either, it goes away when things are run in release mode...

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2009-03-08

Looks like a precision problem somewhere, but it is not clear where. Can you run the following code and try to track down where the nans come from in x_series?

```
import numpy as np

def check_loss_of_precision(dtype):
    """Check loss of precision in complex arc* functions"""

    # Check against known-good functions

    info = np.finfo(dtype)
    real_dtype = dtype(0.).real.dtype
    eps = info.eps

    x_series = np.logspace(np.log10(info.tiny/eps).real, -3, 200, endpoint=False)
    x_basic = np.logspace(dtype(-3.).real, -1e-8, 10)

    print x_series

if __name__ == "__main__" :
    check_loss_of_precision(np.longcomplex)
```

**Comment by thouis** [comment]
Comment in Trac by atmention:WarrenWeckesser, 2009-03-28

With numpy 1.3.0b1, I get the same failure on Mac OSX 10.4.11 PPC:
    ======================================================================
    FAIL: test_umath.TestComplexFunctions.test_loss_of_precision(<type 'numpy.complex256'>,)
    ----------------------------------------------------------------------
    Traceback (most recent call last):
      File "/Users/wweckesser/numpy130b1/lib/python2.5/site-packages/nose-0.10.4-py2.5.egg/nose/case.py", line 182, in runTest
        self.test(_self.arg)
      File "/Users/wweckesser/numpy130b1/lib/python2.5/site-packages/numpy/core/tests/test_umath.py", line 498, in check_loss_of_precision
        check(x_series, 2_eps)
      File "/Users/wweckesser/numpy130b1/lib/python2.5/site-packages/numpy/core/tests/test_umath.py", line 480, in check
        assert np.all(d < rtol), (np.argmax(d), x[np.argmax(d)], d.max())
    AssertionError: (0, nan, nan)

```
----------------------------------------------------------------------
Ran 2029 tests in 23.644s

FAILED (KNOWNFAIL=1, SKIP=1, failures=1)
```

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2009-03-28

This should be fixed in svn and the upcoming 1.3.0rc1.

**Comment by thouis** [comment]
Comment in Trac by atmention:charris, 2009-03-30

Reported fixed in 1.3.0rc1
